### PR TITLE
fix: optimize_read_in_order for nullable keys

### LINKED ISSUE #63460: ORDER BY NULLS FIRST doesn't work correctly with optimize_read_in_order=1 and rows in same part.
**Describe what's wrong**

ORDER BY NULLS FIRST doesn't work correctly with optimize_read_in_order=1 and rows in same part

**How to reproduce**

```
CREATE TABLE users (uid Nullable(Int16), name String, age Int16) ENGINE=MergeTree ORDER BY (uid) SETTINGS allow_nullable_key=1;

INSERT INTO users VALUES (1, 'John', 33);
INSERT INTO users VALUES (2, 'Ksenia', 48);
INSERT INTO users VALUES (NULL, 'Mark', 50);
OPTIMIZE TABLE users FINAL;

SELECT * FROM users ORDER BY uid ASC LIMIT 10;

SELECT '##########################';

SELECT * FROM users ORDER BY uid ASC NULLS FIRST LIMIT 10;

SELECT '##########################';


SELECT * FROM users ORDER BY uid ASC NULLS FIRST SETTINGS optimize_read_in_order=0;


1	John	33
2	Ksenia	48
\N	Mark	50
##########################
1	John	33
2	Ksenia	48
\N	Mark	50
##########################
\N	Mark	50
1	John	33
2	Ksenia	48

```

https://fiddle.clickhouse.com/ec292b92-aa93-4df9-9372-d9aee1a045ba

### PR DESCRIPTION
**Pull request by korowa**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed `optimize_read_in_order` behaviour for ORDER BY ... NULLS FIRST / LAST on tables with nullable keys

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>CI Settings</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step
- [ ]  Allow: Integration Tests
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Unit tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All with aarch64
- [ ]  Allow: All with ASAN
- [ ]  Allow: All with TSAN
- [ ]  Allow: All with Analyzer
- [ ]  Allow: All with Azure
- [ ]  Allow: Add your option here
---
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN
- [ ]  Exclude: All with MSAN
- [ ]  Exclude: All with UBSAN
- [ ]  Exclude: All with Coverage
- [ ]  Exclude: All with Aarch64
---
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)
- [ ]  allow: batch 1 for multi-batch jobs
- [ ]  allow: batch 2
- [ ]  allow: batch 3
- [ ]  allow: batch 4, 5 and 6
</details>

Closes #63460

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 9ed562ee264903a4681ef6098fe623cc2be54320 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64483/9ed562ee264903a4681ef6098fe623cc2be54320/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>⏳ pending</td></tr>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/64483/9ed562ee264903a4681ef6098fe623cc2be54320/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp [hunk 1]
```diff
@@ -371,16 +373,25 @@ InputOrderInfoPtr buildInputOrderInfo(
...
         if (sort_column_description.collator)
             break;
 
+        /// Since sorting key columns are always sorted with NULLS LAST, reading in order
+        /// supported only for ASC NULLS LAST ("in order"), and DESC NULLS FIRST ("reverse")
+        const auto column_is_nullable = sorting_key.data_types[next_sort_key]->isNullable();
+        if (column_is_nullable
+            && !(
+                (sort_column_description.direction == 1 && sort_column_description.nulls_direction == 1)
+                || (sort_column_description.direction == -1 && sort_column_description.nulls_direction == 1)))
```
Comment (jkartseva) [inline src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp:392]: Can this condition be more readable?
```
column_is_nullable && 
sort_column_description.nulls_direction != 1 && 
sort_column_description.direction != sort_column_description.nulls_direction
```

File: tests/queries/0_stateless/03164_optimize_read_in_order_nullable.sql [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+INSERT INTO 03164_users VALUES (1, 'John', 33);
+INSERT INTO 03164_users VALUES (2, 'Ksenia', 48);
+INSERT INTO 03164_users VALUES (NULL, 'Mark', 50);
+OPTIMIZE TABLE 03164_users FINAL;
+
+SELECT '-- Reproducer result:';
+
+SELECT * FROM 03164_users ORDER BY uid ASC NULLS FIRST LIMIT 10 SETTINGS optimize_read_in_order = 1;
+
+DROP TABLE IF EXISTS 03164_users;
```
Comment (jkartseva) [inline tests/queries/0_stateless/03164_optimize_read_in_order_nullable.sql:14]: DROP TABLE IF EXISTS 03164_multi_key;

File: src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp (hunk 1 above)
Comment (korowa) [reply src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp]: Done (null ordering check is enough) -- initially decided to make this check as straightforward as possible.

**Comment by korowa** [comment]
Failed `03040_dynamic_type_alters_2` seems to be irrelevant (no sorting actions there), and passes locally on the branch.

Looks like original column data has been lost after first rename there, while data inserted after the first rename was available after the second one, as expected.

**Review by jkartseva — APPROVED** [review]

**Comment by korowa** [comment]
@jkartseva kind ping -- doesn't look like auto-merge is going to happen.

**Comment by jkartseva** [comment]
@korowa Please rebase or merge master into your branch. If it won't help, I'll see if I can force-merge it.

**Pull request merged**